server/container_create.go

```diff
@@ -87,15 +87,13 @@ func addOCIBindMounts(mountLabel string, containerConfig *pb.ContainerConfig, sp
 		}
 		src := filepath.Join(bindMountPrefix, mount.HostPath)
 
-		if _, err := os.Lstat(src); err != nil && os.IsNotExist(err) {
-			if err1 := os.MkdirAll(src, 0644); err1 != nil {
-				return nil, nil, fmt.Errorf("Failed to mkdir %s: %s", src, err)
-			}
-		}
-
 		src, err := resolveSymbolicLink(src, bindMountPrefix)
 		if err != nil {
-			return nil, nil, fmt.Errorf("failed to resolve symlink %q: %v", src, err)
+			if !os.IsNotExist(err) {
+				return nil, nil, fmt.Errorf("failed to resolve symlink %q: %v", src, err)
```

Comment:
Can we cache the original (`originalSource := src`) before calling `resolveSymlink` and use `originalSource` here and below?  Or maybe something like:

```go
pathContext := src
if src != originalSource {
  pathContext = fmt.Sprintf("%q (resolved from %q)", pathContext, originalSource)
}
```

and then use `pathContext` here and below?

Reply:
something like?

```
diff --git a/server/container_create.go b/server/container_create.go
index ab936589..0f73bd9e 100644
--- a/server/container_create.go
+++ b/server/container_create.go
@@ -87,8 +87,10 @@ func addOCIBindMounts(mountLabel string, containerConfig *pb.ContainerConfig, sp
                }
                src := filepath.Join(bindMountPrefix, mount.HostPath)
 
-               src, err := resolveSymbolicLink(src, bindMountPrefix)
-               if err != nil {
+               resolvedSrc, err := resolveSymbolicLink(src, bindMountPrefix)
+               if err == nil {
+                       src = resolvedSrc
+               } else {
                        if !os.IsNotExist(err) {
                                return nil, nil, fmt.Errorf("failed to resolve symlink %q: %v", src, err)
                        } else if err1 := os.MkdirAll(src, 0644); err1 != nil {
@@ -269,7 +271,7 @@ func addImageVolumes(rootfs string, s *Server, containerInfo *storage.ContainerI
 func resolveSymbolicLink(path, scope string) (string, error) {
        info, err := os.Lstat(path)
        if err != nil {
-               return path, err
+               return "", err
        }
        if info.Mode()&os.ModeSymlink != os.ModeSymlink {
                return path, nil

```

Reply:
> something like? …

I like returning `path` in the `Lstat` error case, because knowing the path that failed may be useful, while getting an empty string back certainly isn't ;).

The `addOCIBindMounts` changes you have in that suggestion look good to me.

Reply:
I amended this change.  I left the path="" in the Lstat error case as `symlink.FollowSymlinkInScope` doesn't return the path in the error case so `resolveSymbolicLink` would have two different behaviours depending where it fails